File: src/stores/subscription.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import { invoke } from "@tauri-apps/api/core";
import type { Subscription, UpdateSubscriptionRequest } from "@/types/mqtt";
import { validateSubscribeTopic } from "@/utils/mqttErrorHandler";
import { useEnvStore } from "@/stores/env";

export const useSubscriptionStore = defineStore("subscription", () => {
  const subscriptions = ref<Map<number, Subscription[]>>(new Map());
  const loading = ref(false);

  const getSubscriptionsByServer = computed(() => {
    return (serverId: number) => subscriptions.value.get(serverId) || [];
  });

  async function fetchSubscriptions(serverId: number) {
    loading.value = true;
    try {
      const result = await invoke<Subscription[]>("get_subscriptions", {
        serverId,
      });
      subscriptions.value.set(serverId, result);
    } finally {
      loading.value = false;
    }
  }

  async function addSubscription(serverId: number, topic: string, qos: number) {
    // 获取环境变量并替换 topic 中的变量
    const envStore = useEnvStore();
    if (envStore.variables.length === 0) {
      await envStore.loadVariables(serverId);
    }
    const processedTopic = envStore.replaceVariables(topic);
    
    // 验证订阅 Topic（使用替换后的 topic）
    const validation = validateSubscribeTopic(processedTopic);
    if (!validation.valid) {
      throw new Error(validation.error || "Topic 格式无效");
    }

    const result = await invoke<Subscription>("add_subscription", {
      serverId,
      topic: processedTopic,
      qos,
    });

    const serverSubs = subscriptions.value.get(serverId) || [];
    serverSubs.unshift(result);
    subscriptions.value.set(serverId, serverSubs);

    return result;
  }

  async function removeSubscription(
    subscriptionId: number,
    serverId: number,
    topic: string
  ) {
    await invoke("remove_subscription", {
      subscriptionId,
      serverId,
      topic,
    });

    const serverSubs = subscriptions.value.get(serverId) || [];
    const index = serverSubs.findIndex((s) => s.id === subscriptionId);
    if (index !== -1) {
      serverSubs.splice(index, 1);
      subscriptions.value.set(serverId, serverSubs);
    }
  }

  async function toggleSubscription(
    subscriptionId: number,
    serverId: number,
    topic: string,
    qos: number,
    isActive: boolean
  ) {
    await invoke("toggle_subscription", {
      subscriptionId,
      serverId,
      topic,
      qos,
      isActive,
    });

    const serverSubs = subscriptions.value.get(serverId) || [];
    const sub = serverSubs.find((s) => s.id === subscriptionId);
    if (sub) {
      sub.is_active = isActive;
    }
  }

  async function updateSubscription(
    serverId: number,
    oldTopic: string,
    request: UpdateSubscriptionRequest
  ) {
    // 如果 topic 改变了，需要验证新 topic
    if (request.topic) {
      const validation = validateSubscribeTopic(request.topic);
      if (!validation.valid) {
        throw new Error(validation.error || "Topic 格式无效");
      }
    }

    const result = await invoke<Subscription>("update_subscription", {
      serverId,
      oldTopic,
      request,
    });

    const serverSubs = subscriptions.value.get(serverId) || [];
    const index = serverSubs.findIndex((s) => s.id === request.id);
    if (index !== -1) {
      serverSubs[index] = result;
      subscriptions.value.set(serverId, serverSubs);
    }

    return result;
  }

  // 根据 topic 获取订阅（用于消息列表查找颜色）
  function getSubscriptionByTopic(serverId: number, topic: string): Subscription | undefined {
    const serverSubs = subscriptions.value.get(serverId) || [];
    // 先精确匹配
    let sub = serverSubs.find((s) => s.topic === topic);
    if (sub) return sub;
    
    // 再模糊匹配（支持通配符）
    for (const s of serverSubs) {
      if (matchTopic(s.topic, topic)) {
        return s;
      }
    }
    return undefined;
  }

  // MQTT topic 通配符匹配
  function matchTopic(pattern: string, topic: string): boolean {
    // # 匹配任意层级
    if (pattern === "#") return true;
    
    const patternParts = pattern.split("/");
    const topicParts = topic.split("/");
    
    for (let i = 0; i < patternParts.length; i++) {
      const p = patternParts[i];
      
      // # 匹配剩余所有层级
      if (p === "#") return true;
      
      // + 匹配单层级
      if (p === "+") {
        if (i >= topicParts.length) return false;
        continue;
      }
      
      // 精确匹配
      if (p !== topicParts[i]) return false;
    }
    
    return patternParts.length === topicParts.length;
  }

  return {
    subscriptions,
    loading,
    getSubscriptionsByServer,
    fetchSubscriptions,
    addSubscription,
    removeSubscription,
    toggleSubscription,
    updateSubscription,
    getSubscriptionByTopic,
  };
});

File: src-tauri/src/commands/subscription.rs
use crate::db::models::{Subscription, UpdateSubscriptionRequest};
use crate::db::Storage;
use crate::mqtt::MqttManager;
use tauri::State;

#[tauri::command]
pub async fn add_subscription(
    storage: State<'_, Storage>,
    mqtt_manager: State<'_, MqttManager>,
    server_id: i64,
    topic: String,
    qos: i32,
) -> Result<Subscription, String> {
    // 创建订阅
    let sub = Subscription {
        id: None,
        server_id,
        topic: topic.clone(),
        qos,
        is_active: true,
        color: None,
        created_at: None,
    };

    let subscription = storage.create_subscription(sub)?;

    // 如果已连接，则订阅主题
    if mqtt_manager.is_connected(server_id) {
        mqtt_manager
            .subscribe(server_id, topic, qos as u8)
            .await
            .map_err(|e| e.to_string())?;
    }

    Ok(subscription)
}

#[tauri::command]
pub async fn remove_subscription(
    storage: State<'_, Storage>,
    mqtt_manager: State<'_, MqttManager>,
    subscription_id: i64,
    server_id: i64,
    topic: String,
) -> Result<(), String> {
    // 如果已连接，则取消订阅
    if mqtt_manager.is_connected(server_id) {
        mqtt_manager
            .unsubscribe(server_id, topic)
            .await
            .map_err(|e| e.to_string())?;
    }

    // 从存储删除
    storage.delete_subscription(subscription_id)
}

#[tauri::command]
pub async fn get_subscriptions(
    storage: State<'_, Storage>,
    server_id: i64,
) -> Result<Vec<Subscription>, String> {
    Ok(storage.get_subscriptions(server_id))
}

#[tauri::command]
pub async fn toggle_subscription(
    storage: State<'_, Storage>,
    mqtt_manager: State<'_, MqttManager>,
    subscription_id: i64,
    server_id: i64,
    topic: String,
    qos: i32,
    is_active: bool,
) -> Result<(), String> {
    // 更新存储状态
    storage.update_subscription_status(subscription_id, is_active)?;

    // 执行订阅/取消订阅操作
    if mqtt_manager.is_connected(server_id) {
        if is_active {
            mqtt_manager
                .subscribe(server_id, topic, qos as u8)
                .await
                .map_err(|e| e.to_string())?;
        } else {
            mqtt_manager
                .unsubscribe(server_id, topic)
                .await
                .map_err(|e| e.to_string())?;
        }
    }

    Ok(())
}

#[tauri::command]
pub async fn update_subscription(
    storage: State<'_, Storage>,
    mqtt_manager: State<'_, MqttManager>,
    server_id: i64,
    old_topic: String,
    request: UpdateSubscriptionRequest,
) -> Result<Subscription, String> {
    let new_topic = request.topic.clone();
    let new_qos = request.qos;

    // 更新存储
    let subscription = storage.update_subscription(request)?;

    // 如果已连接且 topic 或 qos 改变，需要重新订阅
    if mqtt_manager.is_connected(server_id) && subscription.is_active {
        // 如果 topic 改变了，先取消订阅旧的
        if let Some(ref topic) = new_topic {
            if topic != &old_topic {
                mqtt_manager
                    .unsubscribe(server_id, old_topic)
                    .await
                    .map_err(|e| e.to_string())?;
                mqtt_manager
                    .subscribe(server_id, topic.clone(), subscription.qos as u8)
                    .await
                    .map_err(|e| e.to_string())?;
            } else if new_qos.is_some() {
                // topic 没变但 qos 变了，重新订阅
                mqtt_manager
                    .subscribe(server_id, topic.clone(), subscription.qos as u8)
                    .await
                    .map_err(|e| e.to_string())?;
            }
        } else if new_qos.is_some() {
            // 只有 qos 变了
            mqtt_manager
                .subscribe(server_id, subscription.topic.clone(), subscription.qos as u8)
                .await
                .map_err(|e| e.to_string())?;
        }
    }

    Ok(subscription)
}
